### ISSUE TITLE
SimpleTaxProvider

### ISSUE THREAD

**Issue by timosalex** [BroadleafCommerce/CommunityStarter#45]
Hi,

I know this is not an issue but maybe someone here can help me. I am trying to use SimpleTaxProvider wth heat clinic. I am lost with the annotation based bean configuration, so how can i import it into the project?

https://www.broadleafcommerce.com/docs/core/current/broadleaf-concepts/pricing/tax/simple-tax-provider#

thanks!

**Comment by leroymerlin-lrm** [comment]
Maybe something like that will work.  I dit not test it so please be indulgent :-)

```
/**
 * Just a quick snippet
 */
@Configuration
public class LinuxidefixTaxConfig {


    @Merge("blTaxProviders")
    public List<?> myTaxProviders(SimpleTaxProvider blSimpleTaxProvider) {
        //merge
        return Arrays.asList(blSimpleTaxProvider);
    }

    @Bean
    SimpleTaxProvider blSimpleTaxProvider(){
        return new SimpleTaxProvider();
    }
}

```

**Comment by linuxidefix** [comment]
Maybe something like that will work. I dit not test it so please be indulgent :-)

```
/**
 * Just a quick snippet
 */
@Configuration
public class LinuxidefixTaxConfig {


    @Merge("blTaxProviders")
    public List<?> myTaxProviders(SimpleTaxProvider blSimpleTaxProvider) {
        //merge
        return Arrays.asList(blSimpleTaxProvider);
    }

    @Bean
    SimpleTaxProvider blSimpleTaxProvider(){
        return new SimpleTaxProvider();
    }
}
```

**Comment by timosalex** [comment]
@linuxidefix  thanks a lot. I am trying to understand though how to "call" it and provide all the necessary "entry keys". So in your example i've added:

```
 @Bean
    SimpleTaxProvider blSimpleTaxProvider(){
    	SimpleTaxProvider smpl=new SimpleTaxProvider(); 
    	Map<String,Double> map= new HashMap();
    	map.put("SPAIN", 0.08);
    	smpl.setFulfillmentGroupCountryTaxRateMap(map);
        return  smpl;
    }
```

does this bean now called "automatically" from the system every time to calculate to tax, or do i have to add something else?

thanks again!!!

**Comment by linuxidefix** [comment]
This should be automatic.
Your TaxProvider will be registred in the org.broadleafcommerce.core.pricing.service.TaxServiceImpl#providers

So the following method
org.broadleafcommerce.core.pricing.service.TaxServiceImpl#calculateTaxForOrder
which iterates over providers will use the "blSimpleTaxProvider".

**Comment by timosalex** [comment]
you are the boss, it works !!!

just have to study a bit whats the diffrence between these two:

```
smpl.setItemCountryTaxRateMap(map);
smpl.setFulfillmentGroupCountryTaxRateMap(map);
```

thanks again!!

**Comment by timosalex** [comment]
@linuxidefix sorry for posting on a closed issue but trying to understand the behaviour of tax provider. My main question is how can i set the tax to zero. Consider this scenario:

My SimpleTaxProvider looks like this (with just two countries to simplify the example):

 ```
   @Bean
    SimpleTaxProvider blSimpleTaxProvider(){
    	SimpleTaxProvider smpl=new SimpleTaxProvider(); 
    	Map<String,Double> map= new HashMap();
    	map.put("AUSTRIA", 0.24);
    	map.put("BELGIUM", 0.16);
    	smpl.setItemCountryTaxRateMap(map);
        return  smpl;
}
```

If I choose Canada as country (on shippingInfoForm) the system doesn't calculate any tax (which is correct). If I edit my country (go back to edit my shipping address) and choose Austria the system will correctly caulcate a tax of 0,24. If I then go back and choose Belgium it will correctly calculate 0.16. But if i go back and choose a different country (any other country than Austria or Belgium) the tax still remains of 0,16 (or the last value that has been calculated) and is not  set to zero.

In addition if I write this in post method of shipping info controller 

`        	System.out.println(simpletaxProvider.determineItemTaxRate(tempad));
`

I get the correct values. For Austria it prints 0.24, for Belgium 0.16 and for any other country 0.

I am trying to understand this behaviour. It kind of seems like a bug to me. How is possible to set tax to zero or "cancel" it? Shouldn't this be automatic?

thanks again!!!!

**Comment by timosalex** [comment]
actually i find a way to set tax to zero on the scenario above but it seems kind of hacky. after the submission of the shipping address something like that works:

```
    	Order cart = CartState.getCart();
 cart.setTaxOverride(false);
        if(simpletaxProvider.determineItemTaxRate(tempad).doubleValue()==0) {
        	cart.setTaxOverride(true);
    		cart.setTotalTax(new Money());
    		
        }
```
however i feel that there must be a better/automatic solution?

**Comment by cja769** [comment]
@timosalex This seems like a bug in `SimpleTaxProvider` here
```
                        if (factor != null && factor.compareTo(BigDecimal.ZERO) != 0) {
                            TaxDetail tax;
                            checkDetail: {
                                for (TaxDetail detail : fgItem.getTaxes()) {
                                    if (detail.getType().equals(TaxType.COMBINED)) {
                                        tax = detail;
                                        break checkDetail;
                                    }
                                }
                                tax = entityConfig.createEntityInstance(TaxDetail.class.getName(), TaxDetail.class);
                                tax.setType(TaxType.COMBINED);
                                fgItem.getTaxes().add(tax);
                            }
                            tax.setRate(factor);
                            tax.setAmount(fgItem.getTotalItemTaxableAmount().multiply(factor));
                        }
```
in `SimpleTaxProvider#calculateTaxForOrder`. Basically as soon as a tax rate is set it'll never be unset since we skip the core logic. I opened a pull request here https://github.com/BroadleafCommerce/BroadleafCommerce/pull/2081 that fixes the issue. It should be merged tomorrow.

> you are the boss, it works !!!
> 
> just have to study a bit whats the diffrence between these two:
> 
> ```
> smpl.setItemCountryTaxRateMap(map);
> smpl.setFulfillmentGroupCountryTaxRateMap(map);
> ```
> thanks again!!

I'm not sure if you figured this out but the difference between these two is but one taxes at an item level and the other taxes at a fulfillment group level (items being shipped in the same way to the same location). I'm not versed in a lot of the details of taxing but I'm sure different areas have different ways they require you to apply taxes.

**Comment by cja769** [comment]
I merged my change into develop-6.0.x line so you should no longer experience these issues

**Comment by timosalex** [comment]
@cja769 thanks a ton! I am actually still on 5.x version but will check it out soon :+1:
